net/speedtest/speedtest_server.go

```diff
@@ -0,0 +1,159 @@
+// Copyright (c) 2021 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package speedtest
+
+import (
+	"crypto/rand"
+	"encoding/json"
+	"fmt"
+	"net"
+	"time"
+)
+
+// Serve starts up the server on a given host and port pair. It starts to listen for
+// connections and handles each one in a goroutine. Because it runs in an infinite loop,
+// this function only returns if any of the speedtests return with errors, or if a bool is sent
+// to the killSignal channel.
+func Serve(l *net.TCPListener, maxConnections int, killSignal chan bool, resultsChan chan []Result) error {
+	defer l.Close()
+
+	numConnections := 0
+	testStateChan := make(chan TestState, maxConnections)
+	connChan := make(chan *net.TCPConn, maxConnections)
+
+	// This goroutine runs in an infinite loop and returns if there is an error with the listener
+	// or if it is closed. The listener is closed when StartServer returns via a defer statement.
+	go (func() {
+		for {
+			conn, err := l.AcceptTCP()
+			if err != nil {
+				// The AcceptTCP will return an error if the listener is closed.
+				return
```

Comment:
it'd be useful to bubble this up to the outer level, so that if the very first AcceptTCP call fails we can display a useful error message to the user.